crates/sui-core/src/checkpoints/causal_order.rs

```diff
@@ -0,0 +1,421 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use std::{
+    collections::{BTreeMap, BTreeSet, HashSet},
+    sync::Arc,
+};
+
+use sui_types::{
+    base_types::{ExecutionDigests, TransactionDigest},
+    error::SuiResult,
+    messages::TransactionEffects,
+};
+use typed_store::Map;
+
+use crate::authority::AuthorityStore;
+
+use super::CheckpointStore;
+
+/// The interfaces here allow a separation between checkpoint store, that knows about digests largely,
+/// and other parts of the system that know about the transaction semantics, hold and can interpret the
+/// transaction effects.
+///
+/// A point of interaction between these two worlds is necessary when we need to order the execution digests
+/// within a checkpoint, as well as detect digest already in checkpoints of missing to have a full causal history.
+/// The interface here allows these computations to be implemented without passing in a full authority / authority store
+/// for the sake of keeping components separate enough to be tested without one another.
+
+pub trait CausalOrder {
+    fn get_complete_causal_order(
+        &self,
+        transactions: &[ExecutionDigests],
+        ckpt_store: &mut CheckpointStore,
+    ) -> SuiResult<Vec<ExecutionDigests>>;
+}
+
+pub trait EffectsStore {
+    fn get_effects(
+        &self,
+        transactions: &[ExecutionDigests],
+    ) -> SuiResult<Vec<Option<TransactionEffects>>>;
+
+    fn causal_order_from_effects(
+        &self,
+        transactions: &[ExecutionDigests],
+        ckpt_store: &mut CheckpointStore,
+    ) -> SuiResult<Vec<ExecutionDigests>> {
+        let effects = self.get_effects(transactions)?;
+
+        // Ensure all transactions included are executed (static property). This should be true since we should not
+        // be signing a checkpoint unless we have processed all transactions within it.
+        debug_assert!(effects.iter().all(|e| e.is_some()));
+
+        // Include in the checkpoint the computed effects,  rather than the effects provided.
+        // (Which could have been provided by < f+1 byz validators and be incorrect).
```

Comment:
where are these digests coming from that they could be false? Can the digests in a CheckpointContents be false?

Reply:
Good question: a checkpoint is the union of >2/3 proposals, and each proposal contains a set of execution digests (transaction digest + effects digest). So an item in the union may be the input of a single byz nodes in the 2/3 set. Now the transaction cannot be bad because a certificate needs to be included (which is collectivelly signed). But the effects digest is not signed so could be ficticious.

Reply:
let's not store effects digests then - i can't think of any situation where it is ok to read them from a checkpoint given that they could be anything.

Reply:
As we discussed in chat, you can trust the effects in a checkpoint, just not the effects in the proposal. So we keep them in.